### PR TITLE
Add API route to update the group a job belongs to

### PR DESCRIPTION
**Pull request by Martchus**

### CODE REVIEW

**Comment by coolo** [comment]
Sorry, but this is a UPDATE to a job. So it should be a patch route - and then you can replace the current hard coded prio route to support UPDATE specific aspects like prio or group. Also I'm often in need to patch settings of a job to fix up a job before cloning it. This should be part of that route as well instead of adding yet another route per attribute.

**Comment by Martchus** [comment]
Ok.

I chose this approach for consistency because such routes were already present but having a common update route makes sense. (In fact I did that when adding routes for job groups.)

However, moving prio to a common update route would break the API, so I better leave the old version untouched and provide setting it in a common update route only additionally.

And the status update should stay a separate route anyways (because there is a little bit more logic involved than just setting a db value)?

**Comment by coolo** [comment]
Then mark the prio route as deprecate - and support its functionality through update. And yes, the update_status actually sets more than the status, so yes

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm [hunk 1]
```diff
@@ -253,6 +253,30 @@ sub update_status {
...
 }
 
+sub update_group {
+    my ($self) = @_;
+
+    my $job = $self->app->schema->resultset('Jobs')->find(int($self->stash('jobid')));
+    if (!$job) {
+        $self->render(json => {error => 'Job does not exist'}, status => 404);
+        return;
+    }
```
Comment (kraih) [inline lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: Wouldn't this look better here (and for similar cases below)?
```perl
return $self->render(json => {error => 'Job does not exist'}, status => 404)
    unless my $job = $self->app->schema->resultset('Jobs')->find(int($self->stash('jobid')));
```

File: t/api/04-jobs.t [hunk 2]
```diff
@@ -371,4 +371,13 @@ $get = $t->get_ok('/api/v1/jobs', form => $job_properties);
 is($get->tx->res->json->{jobs}->[1]->{group},    'opensuse test');
 is($get->tx->res->json->{jobs}->[1]->{priority}, 42);
 
+$get = $t->get_ok('/api/v1/jobs/99926')->status_is(200);
+is($get->tx->res->json->{job}->{group}, 'opensuse', 'current group is opensuse');
```
Comment (kraih) [inline t/api/04-jobs.t]: There's a test method you can use for this too. http://mojolicious.org/perldoc/Test/Mojo#json_is
```perl
$t->json_is('/job/group' => 'opensuse', 'current group is opensuse');
```

**Comment by kraih** [comment]
Have to agree with @coolo, update operations are best represented with `PATCH' or `PUT` requests.

**Comment by Martchus** [comment]
`POST` is for consistency with the update route for priority (which can not be changed without breaking the API). Bit I included the other suggestions.

**Comment by coolo** [comment]
I explained @Martchus that I won't merge this extra route - and rather wait for him to have time to work on a proper UPDATE route.

File: t/api/04-jobs.t [hunk 3]
```diff
@@ -371,4 +371,13 @@ $get = $t->get_ok('/api/v1/jobs', form => $job_properties);
 is($get->tx->res->json->{jobs}->[1]->{group},    'opensuse test');
 is($get->tx->res->json->{jobs}->[1]->{priority}, 42);
 
+$get = $t->get_ok('/api/v1/jobs/99926')->status_is(200);
```
Comment (kraih) [inline t/api/04-jobs.t]: Those variable assignments (`$get = $t->...`) don't appear to serve a purpose.

File: t/api/04-jobs.t [hunk 4]
```diff
@@ -371,4 +371,13 @@ $get = $t->get_ok('/api/v1/jobs', form => $job_properties);
...
+$t->json_is('/job/group' => 'opensuse', 'current group is opensuse');
```
Comment (kraih) [inline t/api/04-jobs.t]: This line could just be appended to the method chain on the previous line.

**Comment by okurz** [comment]
ok, fine.

@coolo can you give user stories describing what should be supported over the update route? If everything would be allowed to be updated then wouldn't that be dangerous for jobs to set, e.g. currently running jobs.

**Comment by coolo** [comment]
So this is what you get when you get:

```
{
  job => {
           assets     => {
                           iso => [
                                    "openSUSE:42-Staging:Core-Staging-DVD-x86_64-Build57.7-Media.iso",
                                  ],
                         },
           children   => { Chained => [], Parallel => [] },
           clone_id   => undef,
           group      => "openSUSE Leap Staging Projects",
           group_id   => 6,
           id         => 321009,
           name       => "opensuse-42.3:Core-Staging-DVD-x86_64-Build57.7-RAID1\@64bit",
           parents    => { Chained => [], Parallel => [] },
           priority   => 50,
           result     => "passed",
           retry_avbl => 3,
           settings   => {
                           ARCH             => "x86_64",
                           BACKEND          => "qemu",
                           BUILD            => 57.7,
                           DESKTOP          => "minimalx",
                           DISTRI           => "opensuse",
                           FLAVOR           => "Staging-DVD",
                           INSTALLONLY      => 1,
                           ISO              => "openSUSE:42-Staging:Core-Staging-DVD-x86_64-Build57.7-Media.iso",
                           ISO_MAXSIZE      => 4700372992,
                           MACHINE          => "64bit",
                           NAME             => "00321009-opensuse-42.3:Core-Staging-DVD-x86_64-Build57.7-RAID1\@64bit",
                           QEMUCPU          => "qemu64",
                           RAIDLEVEL        => 1,
                           STAGING          => 1,
                           START_AFTER_TEST => "RAID0",
                           TEST             => "RAID1",
                           VERSION          => "42.3:Core",
                           VIRTIO_CONSOLE   => 1,
                           WORKER_CLASS     => "qemu_x86_64",
                         },
           state      => "done",
           t_finished => "2016-12-14T13:05:58",
           t_started  => "2016-12-14T12:50:37",
           test       => "RAID1",
         },
}
```

There are many dangerous properties, but IMO settings, retry_avbl, group/group_id and priority should be set through *one* route.

**Comment by Martchus** [comment]
Should it take JSON or form data?

Form data would be like `priority=60 ARCH=x86_64 group=0 ...`. So job settings and actual columns would be mixed. A typo in a column would like prioriy would not lead to an error, it would create a new entry in the job settings.

In JSON this could be distinguished better. Also there could be a flag like 'replace settings' so only the explicitly specified settings would be present after the update.

However, when using JSON the client must be updated because it currently only allows to post/put form data.

**Comment by coolo** [comment]
@aaannz - in the floor we agreed to use JSON. This is challenging for the client, but IMO we can make it a real openqa client and smart enough to know which route requires which form :)

**Comment by kraih** [comment]
If you wanted to go all out with REST, you could also make it discoverable in the API with `OPTIONS`.
```
OPTIONS /foo/bar HTTP/1.1
Host: example.com

HTTP/1.1 204 No Content
Allow: GET, PATCH
Accept-Patch: application/json
```

**Comment by aaannz** [comment]
I have no problem with sending JSON. Shall I understand this as, eventually, we are going to replace all API form data by JSON?

**Comment by Martchus** [comment]
@aaannz I think we won't replace anything because that would break the API.

New PR providing common route: https://github.com/os-autoinst/openQA/pull/1084/files

**Pull request closed without merging**